Fix HTTP status-code when loading invalid tar

thaJeztah opened the pull request:
fixes https://github.com/moby/moby/issues/34416

Opening for discussion, because this changes the `WriteFlusher` interface, which is used in various places; also the error-code may need to be more specific based on the actual error.

Before this change:

    echo "hello world, I am not a tar" | curl -v --unix-socket  /var/run/docker.sock -H "Content-Type: application/x-tar"  --data-binary @- -X POST http://localhost/v1.29/images/load
    *   Trying /var/run/docker.sock...
    * Connected to localhost (/Users/sebastiaan/Library/Containers/com.dock) port 80 (#0)
    > POST /v1.29/images/load HTTP/1.1
    > Host: localhost
    > User-Agent: curl/7.51.0
    > Accept: */*
    > Content-Type: application/x-tar
    > Content-Length: 28
    >
    * upload completely sent off: 28 out of 28 bytes
    < HTTP/1.1 200 OK
    < Api-Version: 1.31
    < Content-Type: application/json
    < Date: Mon, 07 Aug 2017 15:11:35 GMT
    < Docker-Experimental: true
    < Ostype: linux
    < Server: Docker/17.07.0-ce-rc1 (linux)
    < Transfer-Encoding: chunked
    <
    {"errorDetail":{"message":"Error processing tar file(exit status 1): unexpected EOF"},"error":"Error processing tar file(exit status 1): unexpected EOF"}
    * Curl_http_done: called premature == 0
    * Connection #0 to host localhost left intact

After this change:

    / # echo "hello world, I am not a tar" | curl -v --unix-socket  /var/run/docker.sock -H "Content-Type: application/x-tar"  --data-binary @- -X POST http://localhost/v1.29/images/load
       *   Trying /var/run/docker.sock...
    * Connected to localhost (/var/run/docker.sock) port 80 (#0)
    > POST /v1.29/images/load HTTP/1.1
    > Host: localhost
    > User-Agent: curl/7.52.1
    > Accept: */*
    > Content-Type: application/x-tar
    > Content-Length: 28
    > 
    * upload completely sent off: 28 out of 28 bytes
    < HTTP/1.1 400 Bad Request
    < Api-Version: 1.36
    < Content-Type: application/json
    < Docker-Experimental: false
    < Ostype: linux
    < Server: Docker/dev (linux)
    < Date: Fri, 16 Feb 2018 16:21:46 GMT
    < Transfer-Encoding: chunked
    < 
    Api-Version: 1.36
    Content-Type: application/json
    Docker-Experimental: false
    Ostype: linux
    Server: Docker/dev (linux)
    {"errorDetail":{"message":"Error processing tar file(exit status 1): unexpected EOF"},"error":"Error processing tar file(exit status 1): unexpected EOF"}
    * Curl_http_done: called premature == 0
    * Connection #0 to host localhost left intact

thaJeztah commented:
ping @cpuguy83 could you have a look if this direction is ok or not?

cpuguy83 commented on api/server/router/image/image_routes.go:
```diff
@@ -241,6 +241,7 @@ func (s *imageRouter) postImagesLoad(ctx context.Context, w http.ResponseWriter,
 	output := ioutils.NewWriteFlusher(w)
 	defer output.Close()
 	if err := s.backend.LoadImage(r.Body, output, quiet); err != nil {
+		output.WriteHeader(http.StatusUnprocessableEntity)
```
How are we coming to the conclusion that this is the correct status?
Also I think everywhere else we just use 400 for user errors.

thaJeztah replied:
Yes, was doubting to make it more generic (just a `400`), to prevent having to do string-matching here for other errors.

cpuguy83 commented on api/types/backend/backend.go:
```diff
@@ -64,7 +65,7 @@ type LogSelector struct {
 type ContainerStatsConfig struct {
 	Stream    bool
-	OutStream io.Writer
+	OutStream http.ResponseWriter
```
This is not right.

thaJeztah replied:
I ended up changing it due to https://github.com/moby/moby/blob/22472c8be50ca0ddaf8207d46c6a9f5e792bd0af/daemon/stats.go#L41-L47

(also a reason I opened this for discussion)

cpuguy83 reviewed (changes requested):
Importing HTTP in the backend is not what we want.
In the attach code we pass a function to get the write streams rather than passing the actual writers, which allows the backend to return early due to an error without having sent a 200 response header already.

thaJeztah committed:
Fix HTTP status-code when loading invalid tar

Before this change:

    echo "hello world, I am not a tar" | curl -v --unix-socket  /var/run/docker.sock -H "Content-Type: application/x-tar"  --data-binary @- -X POST http://localhost/v1.29/images/load
    *   Trying /var/run/docker.sock...
    * Connected to localhost (/Users/sebastiaan/Library/Containers/com.dock) port 80 (#0)
    > POST /v1.29/images/load HTTP/1.1
    > Host: localhost
    > User-Agent: curl/7.51.0
    > Accept: */*
    > Content-Type: application/x-tar
    > Content-Length: 28
    >
    * upload completely sent off: 28 out of 28 bytes
    < HTTP/1.1 200 OK
    < Api-Version: 1.31
    < Content-Type: application/json
    < Date: Mon, 07 Aug 2017 15:11:35 GMT
    < Docker-Experimental: true
    < Ostype: linux
    < Server: Docker/17.07.0-ce-rc1 (linux)
    < Transfer-Encoding: chunked
    <
    {"errorDetail":{"message":"Error processing tar file(exit status 1): unexpected EOF"},"error":"Error processing tar file(exit status 1): unexpected EOF"}
    * Curl_http_done: called premature == 0
    * Connection #0 to host localhost left intact

After this change:

    echo "hello world, I am not a tar" | curl -v --unix-socket  /var/run/docker.sock -H "Content-Type: application/x-tar"  --data-binary @- http://localhost/v1.29/images/load
    *   Trying /var/run/docker.sock...
    * Connected to localhost (/var/run/docker.sock) port 80 (#0)
    > POST /v1.29/images/load HTTP/1.1
    > Host: localhost
    > User-Agent: curl/7.55.0
    > Accept: */*
    > Content-Type: application/x-tar
    > Content-Length: 28
    >
    * upload completely sent off: 28 out of 28 bytes
    < HTTP/1.1 400 Bad Request
    < Api-Version: 1.32
    < Content-Type: application/json
    < Docker-Experimental: false
    < Ostype: linux
    < Server: Docker/17.06.0-dev (linux)
    < Date: Thu, 17 Aug 2017 11:14:55 GMT
    < Transfer-Encoding: chunked
    <
    Api-Version: 1.32
    Content-Type: application/json
    Docker-Experimental: false
    Ostype: linux
    Server: Docker/17.06.0-dev (linux)
    {"errorDetail":{"message":"Error processing tar file(exit status 1): unexpected EOF"},"error":"Error processing tar file(exit status 1): unexpected EOF"}

Attempting to import an "export" image using "docker load":

    docker export somecontainer > foo.tar
    cat foo.tar | curl -v --unix-socket  /var/run/docker.sock -H "Content-Type: application/x-tar"  --data-binary @- http://localhost/v1.29/images/load
    *   Trying /var/run/docker.sock...
    * Connected to localhost (/var/run/docker.sock) port 80 (#0)
    > POST /v1.29/images/load HTTP/1.1
    > Host: localhost
    > User-Agent: curl/7.55.0
    > Accept: */*
    > Content-Type: application/x-tar
    > Content-Length: 5910528
    > Expect: 100-continue
    >
    < HTTP/1.1 100 Continue
    * We are completely uploaded and fine
    < HTTP/1.1 400 Bad Request
    < Api-Version: 1.32
    < Content-Type: application/json
    < Docker-Experimental: false
    < Ostype: linux
    < Server: Docker/17.06.0-dev (linux)
    < Date: Thu, 17 Aug 2017 12:34:02 GMT
    < Transfer-Encoding: chunked
    <
    Api-Version: 1.32
    Content-Type: application/json
    Docker-Experimental: false
    Ostype: linux
    Server: Docker/17.06.0-dev (linux)
    {"errorDetail":{"message":"open /var/lib/docker/tmp/docker-import-708284929/bin/json: no such file or directory"},"error":"open /var/lib/docker/tmp/docker-import-708284929/bin/json: no such file or directory"}

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
ping @cpuguy83 @tonistiigi PTAL

cpuguy83 commented on api/server/router/image/image_routes.go:
```diff
@@ -242,6 +242,8 @@ func (s *imageRouter) postImagesLoad(ctx context.Context, w http.ResponseWriter,
 	output := ioutils.NewWriteFlusher(w)
 	defer output.Close()
 	if err := s.backend.LoadImage(r.Body, output, quiet); err != nil {
+		w.WriteHeader(httputils.GetHTTPErrorStatusCode(err))
```
Does this work? Would `LoadImage` have already written to the stream and thus already have sent a `200 OK`?
We could check `output.Flushed()`, although I would pretty much expect `LoadImage` to handle all things on the stream by this point.

cpuguy83 replied:
OK, can we gate this with a check for `!output.Flushed()`?

thaJeztah replied:
Yes, I can add that

cpuguy83 commented on image/tarexport/load.go:
```diff
@@ -77,16 +77,16 @@ func (l *tarexporter) Load(inTar io.ReadCloser, outStream io.Writer, quiet bool)
 		img, err := image.NewFromJSON(config)
 		if err != nil {
-			return err
+			return invalidArchiveError{err}
```
`errdefs.InvalidParameter(err)`

thaJeztah commented:
ping @cpuguy83 updated 👍

thaJeztah commented:
For some reason, the CLI doesn't print the error message with this change; only difference between before/after is the status code:

```patch
diff --git a/before b/after
index 0493ca6..149e900 100644
--- a/before
+++ b/after
@@ -1,13 +1,14 @@
 * upload completely sent off: 28 out of 28 bytes
-< HTTP/1.1 200 OK
+< HTTP/1.1 400 Bad Request
 < Api-Version: 1.36
 < Content-Type: application/json
-< Date: Fri, 16 Feb 2018 16:51:16 GMT
-< Docker-Experimental: true
+< Docker-Experimental: false
 < Ostype: linux
-< Server: Docker/18.02.0-ce (linux)
+< Server: Docker/dev (linux)
+< Date: Fri, 16 Feb 2018 16:50:38 GMT
 < Transfer-Encoding: chunked
 < 
 {"errorDetail":{"message":"Error processing tar file(exit status 1): unexpected EOF"},"error":"Error processing tar file(exit status 1): unexpected EOF"}
+* Curl_http_done: called premature == 0
 * Connection #0 to host localhost left intact
``

But with this change, the actual error is not printed:

```bash
echo "fooasdsddsdas" | docker load
Error response from daemon: 
```

Wondering if it's an issue in the CLI and it stops early on a 400 status?

AkihiroSuda commented:
> Wondering if it's an issue in the CLI and it stops early on a 400 status?

Is it safer to keep the current behavior for ensuring compatibility?

olljanat commented:
Derek add label: rebuild/*

olljanat commented:
@thaJeztah CI looks failing to:
```
18:40:58 These files import internal code: (either directly or indirectly)
18:40:58  - pkg/chrootarchive/archive_unix.go imports github.com/docker/docker/errdefs
18:40:58 
18:40:59 Build step 'Execute shell' marked build as failure
```

olljanat commented:
Derek add label: status/failing-ci

thaJeztah commented:
Hmm.. good call; looks like we have some fixing to do with the ppc64le and s390x builds (they're disabled by default now, but were started on this)

ppc64le

```
[2021-11-12T09:51:47.615Z]  > [criu 2/2] RUN --mount=type=cache,sharing=locked,id=moby-criu-aptlib,target=/var/lib/apt     --mount=type=cache,sharing=locked,id=moby-criu-aptcache,target=/var/cache/apt         echo 'deb https://download.opensuse.org/repositories/devel:/tools:/criu/Debian_11/ /' > /etc/apt/sources.list.d/criu.list         && apt-get update         && apt-get install -y --no-install-recommends criu         && install -D /usr/sbin/criu /build/criu:
[2021-11-12T09:51:47.615Z] #46 7.620 Reading package lists...
[2021-11-12T09:51:47.615Z] #46 12.62 Reading package lists...
[2021-11-12T09:51:47.615Z] #46 17.28 Building dependency tree...
[2021-11-12T09:51:47.615Z] #46 18.13 Reading state information...
[2021-11-12T09:51:47.615Z] #46 18.30 Package criu is not available, but is referred to by another package.
[2021-11-12T09:51:47.615Z] #46 18.30 This may mean that the package is missing, has been obsoleted, or
[2021-11-12T09:51:47.615Z] #46 18.30 is only available from another source
[2021-11-12T09:51:47.615Z] #46 18.30
[2021-11-12T09:51:47.615Z] #46 18.85 E: Package 'criu' has no installation candidate
```

s390x

```
[2021-11-12T09:54:16.489Z] + sudo modprobe ip6table_filter
[2021-11-12T09:54:16.489Z] sudo: a terminal is required to read the password; either use the -S option to read from standard input or configure an askpass helper
script returned exit code 1
```
